packages/ui/src/components/EasySetup/ManualExtrinsic.tsx

```diff
@@ -137,11 +90,84 @@ const ManualExtrinsic = ({
     })
   }, [inputParams, paramFields])
 
+  const transformParams = useCallback(
+    (paramFields: ParamField[], inputParams: any[], opts = { emptyAsNull: true }) => {
+      // if `opts.emptyAsNull` is true, empty param value will be added to res as `null`.
+      // otherwise, it will not be added
+      const paramVal = inputParams.map((inputParam) => {
+        // to cater the js quirk that `null` is a type of `object`.
+        if (
+          typeof inputParam === 'object' &&
+          inputParam !== null &&
+          typeof inputParam.value === 'string'
+        ) {
+          return inputParam.value.trim()
+        } else if (typeof inputParam === 'string') {
+          return inputParam.trim()
+        }
+        return inputParam
+      })
+
+      const params = paramFields.map((field, ind) => ({
+        ...field,
+        value: paramVal[ind] || null
+      }))
+
+      return params.reduce((previousValue, { type = 'string', value, typeName }) => {
+        if (value == null || value === '')
+          return opts.emptyAsNull ? [...previousValue, null] : previousValue
+
+        let converted = value
+
+        // Deal with a vector
+        if (type.indexOf('Vec<') >= 0) {
+          converted = converted.split(',').map((e: string) => e.trim())
+          converted = converted.map((single: any) =>
+            isNumType(type)
+              ? single.indexOf('.') >= 0
+                ? Number.parseFloat(single)
+                : Number.parseInt(single)
+              : single
+          )
+          return [...previousValue, converted]
+        }
+
```

Comment:
It's actually super complex to read and understand. It would be nice to refactor the lines of code by human-readable actions As an example of refactoring such code is easier to understand when the code of each data manipulation is separated by the methods. This way the method gives you understanding of what is suppose to happen without digging into details

```js

const transformParams = useCallback(
  (paramFields, inputParams, opts = { emptyAsNull: true }) => {
    const params = paramFields.map((field, ind) => ({
      ...field,
      value: processParamValue(inputParams[ind], opts.emptyAsNull),
    }));

    return params.reduce((previousValue, { type = 'string', value, typeName }) => {
      if (isEmptyValue(value)) return opts.emptyAsNull ? [...previousValue, null] : previousValue;

      let converted = value;

      if (isVecType(type)) {
        converted = convertVector(value, type);
        return [...previousValue, converted];
      }

      if (isTypeBalance(typeName) && shouldConvertToBalance(chainInfo, converted)) {
        return handleBalanceConversion(previousValue, converted);
      }

      if (isNumType(type)) {
        converted = convertToNumber(converted);
      }

      return [...previousValue, converted];
    }, []);
  },
  [chainInfo, onSetErrorMessage]
);

function processParamValue(inputParam, emptyAsNull) {
  if (typeof inputParam === 'object' && inputParam !== null && typeof inputParam.value === 'string') {
    return inputParam.value.trim();
  } else if (typeof inputParam === 'string') {
    return inputParam.trim();
  }
  return inputParam;
}

function isEmptyValue(value) {
  return value == null || value === '';
}

function isVecType(type) {
  return type.indexOf('Vec<') >= 0;
}

function convertVector(value, type) {
  const converted = value.split(',').map((e) => e.trim());
  return converted.map((single) =>
    isNumType(type) ? (single.indexOf('.') >= 0 ? parseFloat(single) : parseInt(single)) : single
  );
}

function shouldConvertToBalance(chainInfo, converted) {
  if (!chainInfo?.tokenDecimals) return false;
  if (!converted.match(/^[0-9]+([.][0-9]+)?$/)) {
    onSetErrorMessage('Only numbers and "." are accepted.');
    return false;
  }
  const bnResult = inputToBn(chainInfo.tokenDecimals, converted);
  if (bnResult.gte(getGlobalMaxValue(128))) {
    onSetErrorMessage('Amount too large');
    return false;
  }
  return true;
}

function handleBalanceConversion(previousValue, converted) {
  const bnResult = inputToBn(chainInfo.tokenDecimals, converted);
  return [...previousValue, bnResult.toString()];
}

function convertToNumber(converted) {
  return converted.indexOf('.') >= 0 ? parseFloat(converted) : parseInt(converted);
}

export default transformParams;
```